tests/e2e/e2e_test.go

```diff
@@ -17,46 +18,587 @@
+//
+//nolint:unused
+func jobIsCompleted(ctx context.Context, cli clientset.Interface, namespace, podName string) bool {
+	pod, err := cli.CoreV1().Pods(namespace).Get(ctx, podName, metav1.GetOptions{})
+	if err != nil {
+		return false
+	}
+	// Check if the pod's phase is Succeeded.
+	if pod.Status.Phase == "Succeeded" {
+		return true
+	}
+	return false
+}
 
-	ginkgo.RunSpecs(t, "nvidia k8s-device-plugin e2e suite", suiteConfig, reporterConfig)
+// randomSuffix provides a random sequence to append to pods,services,rcs.
+//
+//nolint:unused
+func randomSuffix() string {
+	return strconv.Itoa(rand.Intn(10000))
+}
+
+// getNonControlPlaneNodes gets the nodes that are not tainted for exclusive control-plane usage
+//
+//nolint:unused
+func getNonControlPlaneNodes(ctx context.Context, cli clientset.Interface) ([]corev1.Node, error) {
+	nodeList, err := cli.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	if len(nodeList.Items) == 0 {
+		return nil, fmt.Errorf("no nodes found in the cluster")
+	}
+
+	controlPlaneTaint := corev1.Taint{
+		Effect: corev1.TaintEffectNoSchedule,
+		Key:    "node-role.kubernetes.io/control-plane",
+	}
+	out := []corev1.Node{}
+	for _, node := range nodeList.Items {
+		if !taintExists(node.Spec.Taints, &controlPlaneTaint) {
+			out = append(out, node)
+		}
+	}
+
+	if len(out) == 0 {
+		return nil, fmt.Errorf("no non-control-plane nodes found in the cluster")
+	}
+	return out, nil
+}
+
+// taintExists checks if the given taint exists in list of taints. Returns true if exists false otherwise.
+//
+//nolint:unused
+func taintExists(taints []corev1.Taint, taintToFind *corev1.Taint) bool {
+	for _, taint := range taints {
+		if taint.MatchTaint(taintToFind) {
+			return true
+		}
+	}
+	return false
+}
+
+// getNode returns the node object from the list of nodes
+//
+//nolint:unused
+func getNode(nodes []corev1.Node, nodeName string) corev1.Node {
+	for _, node := range nodes {
+		if node.Name == nodeName {
+			return node
+		}
+	}
+	return corev1.Node{}
+}
+
+// CreateOrUpdateJobsFromFile creates or updates jobs from a file
+func CreateOrUpdateJobsFromFile(ctx context.Context, cli clientset.Interface, filename, namespace string) ([]string, error) {
+	jobs, err := newJobFromfile(filepath.Join(packagePath, "data", filename))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Job from file: %w", err)
+	}
+
+	names := make([]string, len(jobs))
+	for i, job := range jobs {
+		job.Namespace = namespace
+
+		names[i] = job.Name
+
+		// create or update the job
+		_, err := cli.BatchV1().Jobs(namespace).Get(ctx, job.Name, metav1.GetOptions{})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				_, err = cli.BatchV1().Jobs(namespace).Create(ctx, job, metav1.CreateOptions{})
+				if err != nil {
+					return nil, fmt.Errorf("failed to create Job %s: %w", job.Name, err)
+				}
+			} else {
+				return nil, fmt.Errorf("failed to get Job %s: %w", job.Name, err)
+			}
+		}
```

Comment:
Is this nesting patter common in other projects?
We could use:
```suggestion
			if apierrors.IsNotFound(err) {
				if _, cerr := cli.BatchV1().Jobs(namespace).Create(ctx, job, metav1.CreateOptions{}); cerr != nil {
					err = errors.Join(err, fmt.Errorf("failed to create job: %w", cerr))
				}
			}
			if err != nil {
				return nil, fmt.Errorf("failed to get Job %s: %w", job.Name, err)
			}
```